src/main/java/com/pingcap/tikv/policy/RetryPolicy.java

```diff
@@ -46,13 +47,15 @@ private boolean checkNotRecoverableException(Status status) {
     return unrecoverableStatus.contains(status.getCode());
   }
 
-  private void handleFailure(Exception e, String methodName, long millis) {
+  private void handleFailure(Exception e, String methodName, long nextBackMills) {
+    if(nextBackMills == BackOff.STOP) {
+          throw new GrpcException("retry is exhausted.", e);
+    }
     Status status = Status.fromThrowable(e);
     if (checkNotRecoverableException(status)) {
-      logger.error(String.format("Failed to recover from last grpc error calling %s.", methodName));
-      throw new GrpcException(e);
+      throw new GrpcUnrecoverableException(e);
```

Comment:
why do we need this extra exception type? We already have a method testing if unrecoverable.
I would do a simple re-throw.

Reply:
trying to do two things in one PR. probably gonna do it in another PR.